# Model Validation: Cross_val_score in Python

## Cross Validation
Cross-validation is a technique for evaluating ML models by training several ML models on subsets of the available input data and evaluating them on the complementary subset of the data.<br>
<br>
Here are the steps involved in Cros Validation:
- You reserve a sample dataset
- Train the model using the remaining opart of dataset
- Use the reserve sampleof dataset test(Validation) set. This will help you to know the effectiveness in model performance.If your model delivers the positive result on validation data, go ahead with current model. It rocks!

## Why cross validation is better?
Cross-validation is usually the preferred method because it gives your model the opportunity to train on multiple train-test splits. This gives you a better indication of how well your model will perform on unseen data.<br>
**Holdout sets are a great start to model validation. However, using a single train and test set if often not enough. Cross-validation is considered the gold standard when it comes to validating model performance and is almost always used when tuning model hyper-parameters.**

In [1]:
# Load dataset
import numpy as np
import pandas as pd

df = pd.read_csv('/content/Purchased_Dataset.csv')
df.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [5]:
# Extract target and features
X = df[['Age','EstimatedSalary']]
y = df['Purchased']

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

X_train, X_test, y_train, y_test = train_test_split(X,y,random_state =5)

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train,y_train)

y_pred = knn.predict(X_test)
metrics.accuracy_score(y_test,y_pred)

0.77

In [9]:
X_train.head()

Age  EstimatedSalary
382   44           139000
84    30            62000
115   40            57000
328   36           118000
99    28            37000

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

X_train, X_test, y_train, y_test = train_test_split(X,y,random_state =12)

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train,y_train)

y_pred = knn.predict(X_test)
metrics.accuracy_score(y_test,y_pred)

0.73

In [12]:
X_train.head()

Age  EstimatedSalary
92    26            15000
328   36           118000
67    23            82000
254   50            44000
360   43           129000

So we see that if we change random state we find different score and also different training set. If you change random state continously the score also changed for the data.

# Cross_val_score
This method requires four parameters:<br>
- estimator: The model to use
- X: the predictor dataset
- y: the response array
- cv: the number of cross validation splits

If you want to use a different scoring function, you can create a scorer by useing **make_scorer()** method.

In [ ]:
# # Example
# # Load the Methods
# from sklearn.metrics import mean_absolute_error, make_scorer
# # Create a scorer
# mae_scorer = make_scorer(mean_absolute_error)
# # Use the scorer
# cross_val_score(<estimator>, <X>, <y>, cv=5, scoring=mae_scorer)

In [14]:
# import library
from sklearn.model_selection import cross_val_score
# define model
knn = KNeighborsClassifier(n_neighbors=4)
# cross val core
print(cross_val_score(knn, X, y, cv=10, scoring ='accuracy'))

[0.725 0.9   0.9   0.9   0.8   0.7   0.775 0.775 0.75  0.7  ]


Here we find 10 different score for the cv=10. For the first experiment we found score 0.725. Now we calculate the mean of this 10 experiments.

In [16]:
print(cross_val_score(knn, X, y, cv=10, scoring ='accuracy').mean())

0.7925


Finally we found 0.792. That is quiet good!

In [17]:
# Try logistic regression
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()s
print (cross_val_score(logreg, X, y, cv=10, scoring = 'accuracy').mean())

0.6425000000000001



Here KNeighborsClassifier is better than Logistic Regression!